pkg/component/etcd/statefulset/statefulset.go

```diff
@@ -476,6 +476,16 @@ func getStsAnnotations(val *Values, sts *appsv1.StatefulSet) map[string]string {
 		},
 		val.Annotations,
 	)
+
+	// If `scaleToMultiNodeAnnotationKey` annotation is already present in etcd statefulset
+	// then it is better to check scale-up was successful or not before removing `scaleToMultiNodeAnnotationKey` annotation.
+	if metav1.HasAnnotation(sts.ObjectMeta, scaleToMultiNodeAnnotationKey) {
+		if (sts.Spec.Replicas != nil && sts.Status.UpdatedReplicas < *sts.Spec.Replicas) || sts.Status.Replicas > sts.Status.UpdatedReplicas {
+			annotations[scaleToMultiNodeAnnotationKey] = ""
+			return annotations
+		}
+	}
+
```

Comment:
``` go
// generateStsAnnotations merges the provided annotations with default annotations and determines if the scaleToMultiNodeAnnotationKey should be added.
func generateStsAnnotations(val *Values, sts *appsv1.StatefulSet) map[string]string {
	// Merge the provided annotations with the default ones.
	annotations := utils.MergeStringMaps(
		map[string]string{
			common.GardenerOwnedBy:   fmt.Sprintf("%s/%s", val.Namespace, val.Name),
			common.GardenerOwnerType: "etcd",
		},
		val.Annotations,
	)

	// Determine if the scaleToMultiNodeAnnotationKey should be added to the annotations.
	shouldAddAnnotation := false
	if sts != nil && sts.Spec.Replicas != nil {
		// If the scaleToMultiNodeAnnotationKey is already present and all pods are not ready,
		// keep the annotation to indicate that scale-up is in progress.
		if metav1.HasAnnotation(sts.ObjectMeta, scaleToMultiNodeAnnotationKey) &&
			((sts.Status.UpdatedReplicas < *sts.Spec.Replicas) || (sts.Status.Replicas > sts.Status.UpdatedReplicas)) {
			shouldAddAnnotation = true
		} else if val.Replicas > 1 && *sts.Spec.Replicas == 1 {
			// If the cluster has been scaled up to multi-node, add the annotation.
			shouldAddAnnotation = true
		}
	} else if val.Replicas > 1 && val.StatusReplicas == 1 {
		// If the StatefulSet is not available and the cluster has been scaled up to multi-node, add the annotation.
		shouldAddAnnotation = true
	}

	// Add the scaleToMultiNodeAnnotationKey if necessary.
	if shouldAddAnnotation {
		annotations[scaleToMultiNodeAnnotationKey] = ""
	}

	return annotations
}
```

I would like to propose the following changes:
- Remove the `clusterScaledUpToMultiNode` function and inline the conditions for better readability, which also helps avoid duplicate checks.
- Rename the function to `generateStsAnnotations` for improved clarity and to convey the purpose of the function more accurately.
- Update the comments to provide more context and better explanations for the logic behind adding or keeping the `scaleToMultiNodeAnnotationKey` annotation.

I understand that you didn't modify the other parts of the code, but I believe these changes would enhance readability and maintainability. 
What do you think?

Reply:
As it's very urgent to push these PRs, I would suggest not to introduce any changes even for readability and maintainability. We can raise another PR to do so